### PR TITLE
feat: use zstd compression when writing parquet files

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:
Closes https://github.com/influxdata/influxdb_iox/issues/2190

parquet file size is reduced significantly when a final block level compression is added. See https://github.com/influxdata/influxdb_iox/issues/2190#issuecomment-894385130 for details

# Changes
1. Enable zstd compression when writing parquet files in IOx

# Manual Verification
In addition to the unit test, I also checked that zstd was used when I made some parquet files locally:

To test, run iox like this
```shell
cd /Users/alamb/Software/influxdb_iox && cargo run -- run -v --object-store=file --data-dir=$HOME/.influxdb_iox --server-id=42
```

Make a database:
```shell
cd /Users/alamb/Software/influxdb_iox
./target/debug/influxdb_iox database create parquet_db
./target/debug/influxdb_iox database write parquet_db tests/fixtures/lineproto/metrics.lp
```

Now, edit its rules `./scripts/edit_db_rules  localhost:8082 parquet_db` so it persists immediately:

```
{
  "rules": {
    "name": "parquet_db",
    "partitionTemplate": {
      "parts": [
        {
          "time": "%Y-%m-%d %H:00:00"
        }
      ]
    },
    "lifecycleRules": {
      "bufferSizeSoft": "52428800",
      "bufferSizeHard": "104857600",
      "dropNonPersisted": true,
      "immutable": false,
      "persist": true,
      "workerBackoffMillis": "1000",
      "catalogTransactionsUntilCheckpoint": "100",
      "lateArriveWindowSeconds": 1,
      "persistRowThreshold": "1",
      "persistAgeThresholdSeconds": 1,
      "mubRowThreshold": "1",
      "parquetCacheLimit": "0",
      "maxActiveCompactionsCpuFraction": 1
    },
    "workerCleanupAvgSleep": "500s"
  }
}
```

Then examine the contents of the parquet files using `parquet-tools`:

```shell
alamb@ip-10-0-0-124 influxdb_iox % parquet-tools meta /Users/alamb/.influxdb_iox/42/parquet_db/data/disk/2020-06-11\ 16\:00\:00/1.4b1a7805-d6de-495e-844b-32fa452147c7.parquet
<16\:00\:00/1.4b1a7805-d6de-495e-844b-32fa452147c7.parquet
file:         file:/Users/alamb/.influxdb_iox/42/parquet_db/data/disk/2020-06-11 16:00:00/1.4b1a7805-d6de-495e-844b-32fa452147c7.parquet
creator:      parquet-rs version 5.1.0
extra:        ARROW:schema = /////8wHAAAQAAAAAAAKAA4ADAALAAQACgAAABQAAAAAAAABBAAKAAwAAAAIAAQACgAAAAgAAAAIAAAAAAAAAA0AAADoBgAATAYAAKwFAAAcBQAAjAQAAPgDAABoAwAA2AIAAEgCAAC4AQAAHAEAAJAAAAAEAAAA9vn//zQAAAAQAAAAFAAAAAAAAQMQAAAAUv7//wAAAgAAAAAADAAAAHVzZWRfcGVyY2VudAAAAAABAAAABAAAACz5//8sAAAABAAAAB4AAABpb3g6OmNvbHVtbl90eXBlOjpmaWVsZDo6ZmxvYXQAABEAAABpb3g6OmNvbHVtbjo6dHlwZQAAAH76//8wAAAAEAAAABgAAAAAAAECFAAAAMD5//9AAAAAAAAAAQAAAAAEAAAAdXNlZAAAAAABAAAABAAAALD5//8wAAAABAAAACAAAABpb3g6OmNvbHVtbl90eXBlOjpmaWVsZDo6aW50ZWdlcgAAAAARAAAAaW94Ojpjb2x1bW46OnR5cGUAAAAG+///MAAAABAAAAAYAAAAAAABAhQAAABI+v//QAAAAAAAAAEAAAAABQAAAHRvdGFsAAAAAQAAAAQAAAA4+v//MAAAAAQAAAAgAAAAaW94Ojpjb2x1bW5fdHlwZTo6ZmllbGQ6OmludGVnZXIAAAAAEQAAAGlveDo6Y29sdW1uOjp0eXBlABIAGAAUAAAAEwAIAAAADAAEABIAAAA8AAAAHAAAAAwAAAAAAAAKIAAAAAAAAAAAAAYACAAGAAYAAAAAAAMAAAAAAAAAAAAEAAAAdGltZQAAAAABAAAABAAAANz6//8oAAAABAAAABsAAABpb3g6OmNvbHVtbl90eXBlOjp0aW1lc3RhbXAAEQAAAGlveDo6Y29sdW1uOjp0eXBlAAAAkvv//0AAAAAsAAAAJAAAAAAAAQUIAAAAIAAAAID7//8EAAAAePv//yAAAAAAAAABAAAAAHD7//8EAAAAcGF0aAAAAAABAAAABAAAAGz7//8kAAAABAAAABUAAABpb3g6OmNvbHVtbl90eXBlOjp0YWcAAAARAAAAaW94Ojpjb2x1bW46OnR5cGUAAAAe/P//QAAAACwAAAAkAAAAAAABBQgAAAAgAAAADPz//wQAAAAE/P//IAAAAAAAAAEAAAAA/Pv//wQAAABtb2RlAAAAAAEAAAAEAAAA+Pv//yQAAAAEAAAAFQAAAGlveDo6Y29sdW1uX3R5cGU6OnRhZwAAABEAAABpb3g6OmNvbHVtbjo6dHlwZQAAAEL9//80AAAAEAAAABgAAAAAAAECFAAAAIT8//9AAAAAAAAAAQAAAAALAAAAaW5vZGVzX3VzZWQAAQAAAAQAAAB4/P//MAAAAAQAAAAgAAAAaW94Ojpjb2x1bW5fdHlwZTo6ZmllbGQ6OmludGVnZXIAAAAAEQAAAGlveDo6Y29sdW1uOjp0eXBlAAAAzv3//zgAAAAQAAAAGAAAAAAAAQIUAAAAEP3//0AAAAAAAAABAAAAAAwAAABpbm9kZXNfdG90YWwAAAAAAQAAAAQAAAAI/f//MAAAAAQAAAAgAAAAaW94Ojpjb2x1bW5fdHlwZTo6ZmllbGQ6OmludGVnZXIAAAAAEQAAAGlveDo6Y29sdW1uOjp0eXBlAAAAXv7//zQAAAAQAAAAGAAAAAAAAQIUAAAAoP3//0AAAAAAAAABAAAAAAsAAABpbm9kZXNfZnJlZQABAAAABAAAAJT9//8wAAAABAAAACAAAABpb3g6OmNvbHVtbl90eXBlOjpmaWVsZDo6aW50ZWdlcgAAAAARAAAAaW94Ojpjb2x1bW46OnR5cGUAAABS/v//QAAAACwAAAAkAAAAAAABBQgAAAAgAAAAQP7//wQAAAA4/v//IAAAAAAAAAEAAAAAMP7//wQAAABob3N0AAAAAAEAAAAEAAAALP7//yQAAAAEAAAAFQAAAGlveDo6Y29sdW1uX3R5cGU6OnRhZwAAABEAAABpb3g6OmNvbHVtbjo6dHlwZQAAAN7+//9AAAAALAAAACQAAAAAAAEFCAAAACAAAADM/v//BAAAAMT+//8gAAAAAAAAAQAAAAC8/v//BgAAAGZzdHlwZQAAAQAAAAQAAAC4/v//JAAAAAQAAAAVAAAAaW94Ojpjb2x1bW5fdHlwZTo6dGFnAAAAEQAAAGlveDo6Y29sdW1uOjp0eXBlABIAGAAUABIAEwAIAAAADAAEABIAAAAwAAAAEAAAABgAAAAAAAECFAAAAFT///9AAAAAAAAAAQAAAAAEAAAAZnJlZQAAAAABAAAABAAAAET///8wAAAABAAAACAAAABpb3g6OmNvbHVtbl90eXBlOjpmaWVsZDo6aW50ZWdlcgAAAAARAAAAaW94Ojpjb2x1bW46OnR5cGUAEgAcABgAEgATAAgAFAAMAAQAEgAAAFQAAABAAAAANAAAAAAAAQUQAAAANAAAAAgACAAAAAQACAAAAAwAAAAIAAwABAALAAgAAAAgAAAAAAAAAQAAAAAEAAQABAAAAAYAAABkZXZpY2UAAAEAAAAMAAAACAAMAAQACAAIAAAAJAAAAAQAAAAVAAAAaW94Ojpjb2x1bW5fdHlwZTo6dGFnAAAAEQAAAGlveDo6Y29sdW1uOjp0eXBlAAAA
extra:        IOX:metadata = CAUSDglocw1hAAAAABUYleE0GgRkaXNrIhMyMDIwLTA2LTExIDE2OjAwOjAwKAEyEBIOCbJh4l4AAAAAFQEAAAA6AEIOCaByDWEAAAAAFXA3LxpKDgmgcg1hAAAAABVwNy8a

file schema:  arrow_schema
--------------------------------------------------------------------------------------------------------------------
device:       OPTIONAL BINARY L:STRING R:0 D:1
free:         OPTIONAL INT64 R:0 D:1
fstype:       OPTIONAL BINARY L:STRING R:0 D:1
host:         OPTIONAL BINARY L:STRING R:0 D:1
inodes_free:  OPTIONAL INT64 R:0 D:1
inodes_total: OPTIONAL INT64 R:0 D:1
inodes_used:  OPTIONAL INT64 R:0 D:1
mode:         OPTIONAL BINARY L:STRING R:0 D:1
path:         OPTIONAL BINARY L:STRING R:0 D:1
time:         REQUIRED INT64 L:TIMESTAMP(NANOS,false) R:0 D:0
total:        OPTIONAL INT64 R:0 D:1
used:         OPTIONAL INT64 R:0 D:1
used_percent: OPTIONAL DOUBLE R:0 D:1

row group 1:  RC:45 TS:1886 OFFSET:4
--------------------------------------------------------------------------------------------------------------------
device:        BINARY ZSTD DO:4 FPO:48 SZ:103/97/0.94 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: disk1s1, max: disk1s5, num_nulls not defined]
free:          INT64 ZSTD DO:163 FPO:237 SZ:154/174/1.13 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 882940715008, max: 882941370368, num_nulls not defined]
fstype:        BINARY ZSTD DO:375 FPO:406 SZ:80/62/0.78 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: apfs, max: apfs, num_nulls not defined]
host:          BINARY ZSTD DO:506 FPO:564 SZ:161/143/0.89 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: Andrews-MBP.hsd1.ma.comcast.net, max: Andrews-MBP.hsd1.ma.comcast.net, num_nulls not defined]
inodes_free:   INT64 ZSTD DO:771 FPO:828 SZ:123/111/0.90 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 9766246316, max: 9767978157, num_nulls not defined]
inodes_total:  INT64 ZSTD DO:959 FPO:990 SZ:88/70/0.80 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 9767978160, max: 9767978160, num_nulls not defined]
inodes_used:   INT64 ZSTD DO:1113 FPO:1170 SZ:123/111/0.90 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 3, max: 1731844, num_nulls not defined]
mode:          BINARY ZSTD DO:1301 FPO:1336 SZ:82/64/0.78 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: ro, max: rw, num_nulls not defined]
path:          BINARY ZSTD DO:1429 FPO:1500 SZ:137/119/0.87 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: /, max: /System/Volumes/Data, num_nulls not defined]
time:          INT64 ZSTD DO:1629 FPO:1767 SZ:212/201/0.95 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 2020-06-11T16:51:50.000000000, max: 2020-06-11T16:54:10.000000000, num_nulls not defined]
total:         INT64 ZSTD DO:1899 FPO:1930 SZ:88/70/0.80 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 1000240963584, max: 1000240963584, num_nulls not defined]
used:          INT64 ZSTD DO:2046 FPO:2132 SZ:154/179/1.16 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 4387262464, max: 101144281088, num_nulls not defined]
used_percent:  DOUBLE ZSTD DO:2258 FPO:2547 SZ:381/363/0.95 VC:45 ENC:RLE_DICTIONARY,PLAIN,RLE ST:[min: 0.4944349028834563, max: 10.27800255966235, num_nulls not defined]
alamb@ip-10-0-0-124 influxdb_iox %
```

### CODE REVIEW

File: parquet_file/Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,6 @@ internal_types = {path = "../internal_types"}
 metrics = { path = "../metrics" }
 object_store = {path = "../object_store"}
 observability_deps = { path = "../observability_deps" }
-# Turn off the "arrow" feature; it currently has a bug that causes the crate to rebuild every time
```
Comment (alamb) [inline parquet_file/Cargo.toml:21]: this comment is no longer correct as we are not disabling any features in arrow

**Review by e-dard — APPROVED** [review]

File: parquet_file/src/storage.rs [hunk 2]
```diff
@@ -191,6 +192,16 @@ impl Storage {
...
 
+    fn writer_props(metadata_bytes: &[u8]) -> WriterProperties {
+        WriterProperties::builder()
+            .set_key_value_metadata(Some(vec![KeyValue {
+                key: METADATA_KEY.to_string(),
+                value: Some(base64::encode(&metadata_bytes)),
+            }]))
+            .set_compression(Compression::ZSTD)
+            .build()
+    }
```
Comment (NGA-TRAN) [inline parquet_file/src/storage.rs:203]: Nice refactoring

**Review by NGA-TRAN — APPROVED** [review]
I like your description to manually verify it. I have saved those for my own check in the future. Maybe convert that in a the Doc folder for everyone?

**Comment by alamb** [comment]
> I like your description to manually verify it. I have saved those for my own check in the future. Maybe convert that in a the Doc folder for everyone?

Good idea @NGA-TRAN  -- I took a shot at doing so in https://github.com/influxdata/influxdb_iox/pull/2220

**Pull request merged**